document for intro to problem-solving exercises (#606)
Ada-Developers-Academy/textbook-curriculum: 00-programming-fundamentals/about-problem-solving-exercises.md

- For question #3, your two sub-problems that you list do not need to make up the entire problem.
- For question #5, you are free to write down code that you are seeing from your previous notes, projects, Internet, etc.
- Answer question #5 last. If you need to leave an answer blank, let it be #5.
- These are challenging exercises!
- If your answers from last week seem similar to your answers for this week, that is okay.

## Example Problem-Solving Exercise: Ratings

Given this problem prompt:

> You are given an array of hashes called `restaurants`, where each hash represents the data associated with a restaurant. Each restaurant has a rating. Write code that finds the restaurant that has the highest rating.

Answer the following prompts:

1. List at least 5 questions whose answers would clarify the problem statement
2. List 2 sets of example inputs and their expected output
3. Divide the project prompt into at least 2 different sub-problems.
4. Pick one of those problems from #3 and write a numbered list explaining how to solve that problem without using full lines of code
5. Write down on paper a solution for your answer in #4 in code

### Example Answers

> 1. At least 5 clarifying questions

Reviewer: Can these also be numbered to make referencing against prompts easier?

- The answer is wrong if there aren't 6 questions
- The answer is wrong if there were no questions clarifying the structure of the restaurant hash (what are the keys in each hash, etc.)

#### Reviewing #2 (2 example input/output):

- The answer is wrong if there aren't 2 examples
- The answer is wrong if either of them aren't valid/consistent with each other
- The answer is wrong if the inputs were not arrays, even if it had one or zero restaurants

#### Reviewing #3 (2 sub-problems):

- The answer is wrong if there aren't at least 2 sub-problems listed
- The answer is wrong if any of the sub-problems aren't relevant to the original problem

#### Reviewing #4 (List of steps to solve one sub-problem):

- The answer is wrong if it doesn't have at least two steps
- The answer is wrong if it doesn't make sense

#### Reviewing #5 (Code for solving one sub-problem):

The point of this exercise is not fully check if the code is correct at the moment, but to let them practice writing code with their hands.

- The answer is wrong if it is blank

> - The answer is wrong if it is obviously incorrect code (ie they wrote `puts "Oops!"` for their entire response)

Reviewer: This is student facing:
```suggestion
- The answer is wrong if it is obviously incorrect code (ie you wrote `puts "Oops!"` for your entire response)
```
